### ISSUE TITLE
3DS 2.0 archive builds (w/ Carthage) fail with error: "Failed to verify bitcode in CardinalMobile.framework"

### ISSUE THREAD

**Issue by drhaynes-goahead** [braintree/braintree_ios#460]
### General information

* SDK/Library version: 

`3ds-2-beta.2` branch.

* Environment: 

Distribution builds using Carthage.

* iOS Version and Device: 

Building with Xcode 10.2 on macOS Mojave.

* Integration type and version: 

Carthage 0.33.0

### Issue description

When performing a distribution build ('archive' configuration in Xcode), exporting fails on bitcode verification in CardinalMobile.framework. It looks like it contains i386 symbols in the binary.

Presumably there should be a step to strip these out of the CardinalMobile framework. We're currently using Carthage for builds. I'm thinking this should be handled by the `copy-frameworks` script step of Carthage.

But, adding `$(SRCROOT)/Carthage/Build/iOS/BraintreePaymentFlow.framework/CardinalMobile.framework` to Carthage `copy-frameworks` input file list, and `$(BUILT_PRODUCTS_DIR)/$(FRAMEWORKS_FOLDER_PATH)/CardinalMobile.framework` to the output files list does not fix the issue.

Is this possibly an issue in the bitcode configuration in the way the upstream `CardinalMobile.framework` has been built?

Build output logs:

❌  error: exportArchive: Failed to verify bitcode in CardinalMobile.framework/CardinalMobile:

❌  error: Cannot extract bundle from /var/folders/6q/wgy6jtp12w5gzgm9lzcglpqw0000gn/T/XcodeDistPipeline.~~~pEygPm/Root/Payload/MyApp.app/Frameworks/BraintreePaymentFlow.framework/CardinalMobile.framework/CardinalMobile (i386)

❌  error: Cannot extract bundle from /var/folders/6q/wgy6jtp12w5gzgm9lzcglpqw0000gn/T/XcodeDistPipeline.~~~pEygPm/Root/Payload/MyApp.app/Frameworks/BraintreePaymentFlow.framework/CardinalMobile.framework/CardinalMobile (i386)

❌  error: Cannot extract bundle from /var/folders/6q/wgy6jtp12w5gzgm9lzcglpqw0000gn/T/XcodeDistPipeline.~~~pEygPm/Root/Payload/MyApp.app/Frameworks/BraintreePaymentFlow.framework/CardinalMobile.framework/CardinalMobile (i386)

**Comment by billwerges** [comment]
Hi @drhaynes-goahead,

It looks like `CardinalMobile.framework` doesn't currently support Bitcode. We will look into getting this corrected.

**Comment by drhaynes-goahead** [comment]
Thanks @billwerges any progress on this?

**Comment by billwerges** [comment]
Hi @drhaynes-goahead,

We are working on an update for the beta branch that will correct this issue for Carthage users. If you need to create a build sooner, you can manually remove the x86 slices from the CardinalMobile framework before performing the archive action. Inside of the `CardinalMobile.framework` directory you can run
```
lipo -extract armv7 CardinalMobile -o CardinalMobile-armv7
lipo -extract arm64 CardinalMobile -o CardinalMobile-arm64
lipo -o CardinalMobile -create CardinalMobile-armv7 CardinalMobile-arm64
rm CardinalMobile-armv7
rm CardinalMobile-arm64
```
This will replace the universal binary with one that only has arm architectures. 

After archiving, you will want to revert these changes so the SDK can continue to be used in the simulator.

You can reference this post for more details: https://stackoverflow.com/questions/46404236/failed-to-verify-bitcode-while-exporting-archive-for-ad-hoc-distribution-tried

**Comment by drhaynes-goahead** [comment]
Great, thanks @billwerges - will give that a try in thew meantime.

**Comment by drhaynes-goahead** [comment]
Hmm, after doing that and re-archiving, we're still seeing the same issue when extracting from the .xcarchive:

```
Failed to verify bitcode in CardinalMobile.framework/CardinalMobile:
error: Cannot extract bundle from /var/folders/wn/ztgs3sq17sd6_5l2m135rks80000gn/T/XcodeDistPipeline.~~~q1qkYH/Root/Payload/MyApp.app/Frameworks/BraintreePaymentFlow.framework/Frameworks/CardinalMobile.framework/CardinalMobile (i386)
```

Our Carthage input file for this is:
`$(SRCROOT)/Carthage/Build/iOS/BraintreePaymentFlow.framework/CardinalMobile.framework`

Checking the results of extracting only armv7 and arm64 slices, all seems ok:
```
$ lipo -info Carthage/Build/iOS/BraintreePaymentFlow.framework/CardinalMobile.framework/CardinalMobile
Architectures in the fat file: Carthage/Build/iOS/BraintreePaymentFlow.framework/CardinalMobile.framework/CardinalMobile are: armv7 arm64
```

So the slices now seem ok.

I have deleted derived data, but it seems that the i386 slice is still being picked up from somewhere.

Shouldn't this stripping already be happening by virtue of the Carthage `copy-frameworks` step as part of the normal build process anyway?

**Comment by drhaynes-goahead** [comment]
Any update on this @billwerges ?

**Comment by bull-xu** [comment]
Once update to 4.25.0, the CardinalMobile.framework caused another issue:

```

Undefined symbols for architecture armv7:
  "_lprofDirMode", referenced from:
    -exported_symbol[s_list] command line option
  "_lprofCurFilename", referenced from:
    -exported_symbol[s_list] command line option

...

error: Failed to compile bundle: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/CardinalMobileWvjf3z/CardinalMobile.armv7.xar

[error]Error: /usr/bin/xcodebuild failed with return code: 70

```

Xcode 10.2.1 archive.

**Comment by sestevens** [comment]
Hi @bull-xu, I wasn't able to replicate this issue.

A few questions for you:

- Are you using Carthage? If so, did you add the Cardinal Mobile framework to your Xcode project by following [these instructions](https://developers.braintreepayments.com/guides/3d-secure/migration/ios/v4#additional-setup-for-carthage)?
- Is the error occurring during the build phase when you try to archive your app?

**Comment by bull-xu** [comment]
Hi @sestevens thanks for the speedy response.

Here's my answers to your:

- I'm using Pods, not Carthage.
- I'm using AppCenter for build. The Xcode version is 10.2.1. Braintree version is 4.25.0. Deployment target iOS 10.0. Bitcode is enabled.

Once I downgraded the Braintree version to 4.23.2 and the error was gone. 

UPDATED:
Tried to do the Archive from Xcode locally by using the 4.25.0 and there is no error.

**Comment by sestevens** [comment]
@bull-xu Great, thanks for letting us know!

**Comment by billwerges** [comment]
Hi @drhaynes-goahead, using the most recent release and following [these instructions](https://developers.braintreepayments.com/guides/3d-secure/migration/ios/v4#additional-setup-for-carthage), you should be able to properly create an archive build now. Let us know if you are still running into any issues.

**Comment by bull-xu** [comment]
Hi @sestevens 

Sorry to bring the issue live again due to the build on AppCenter failed today. If you don't mind, would you please take a look into the logs below? Thanks a lot.

```

    "/Applications/Xcode_10.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" "-arch" "armv7" "-dylib" "-compatibility_version" "1.0.0" "-current_version" "1.0.0.0.0" "-install_name" "@rpath/CardinalMobile.framework/CardinalMobile" "-ios_version_min" "9.0.0" "-rpath" "@executable_path/Frameworks" "-rpath" "@loader_path/Frameworks" "-dead_strip" "-exported_symbols_list" "exports.exp" "-syslibroot" "/Applications/Xcode_10.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk" "-sdk_version" "12.2.0" "-filelist" "/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/CardinalMobile5Mk4co/CardinalMobile.armv7.out.LinkFileList" "-ignore_auto_link" "-allow_dead_duplicates" "/Applications/Xcode_10.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk/System/Library/Frameworks/Foundation.framework/Foundation.tbd" "/Applications/Xcode_10.2.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS12.2.sdk/usr...
    -= Output =-
    Undefined symbols for architecture armv7:
      "_lprofDirMode", referenced from:
         -exported_symbol[s_list] command line option
      "_lprofCurFilename", referenced from:
         -exported_symbol[s_list] command line option
      "___llvm_profile_raw_version", referenced from:
         -exported_symbol[s_list] command line option
      "_lprofMergeValueProfData", referenced from:
         -exported_symbol[s_list] command line option
      "___llvm_profile_filename", referenced from:
         -exported_symbol[s_list] command line option
      "___llvm_profile_runtime", referenced from:
          __hidden#3_ in 01.o
          __hidden#3_ in 02.o
          __hidden#3_ in 03.o
          __hidden#3_ in 04.o
          __hidden#3_ in 05.o
          __hidden#3_ in 06.o
          __hidden#3_ in 07.o
          ...
    ld: symbol(s) not found for architecture armv7
    Exited with 1
    
    
    error: Failed to compile bundle: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/CardinalMobile5Mk4co/CardinalMobile.armv7.xar
    

Stderr:

>
    /Applications/Xcode_10.2.1.app/Contents/Developer/usr/bin/ipatool:246:in `run'
    /Applications/Xcode_10.2.1.app/Contents/Developer/usr/bin/ipatool:2469:in `block in CompileOrStripBitcodeInBundle'
    /Applications/Xcode_10.2.1.app/Contents/Developer/usr/bin/ipatool:2408:in `each'
    /Applications/Xcode_10.2.1.app/Contents/Developer/usr/bin/ipatool:2408:in `CompileOrStripBitcodeInBundle'
    /Applications/Xcode_10.2.1.app/Contents/Developer/usr/bin/ipatool:2632:in `ProcessIPA'
    /Applications/Xcode_10.2.1.app/Contents/Developer/usr/bin/ipatool:3459:in `<main>'}

** EXPORT FAILED **

##[error]Error: /usr/bin/xcodebuild failed with return code: 70

```

**Comment by sestevens** [comment]
@bull-xu Could you open a new GitHub issue for this? The original problem in this issue is specific to Carthage, so I think we should treat the AppCenter build failures you're experiencing as a separate issue.

**Comment by bull-xu** [comment]
> @bull-xu Could you open a new GitHub issue for this? The original problem in this issue is specific to Carthage, so I think we should treat the AppCenter build failures you're experiencing as a separate issue.

Hi @sestevens, I open a new thread for the issue: https://github.com/braintree/braintree_ios/issues/467 And thanks again for your reminding.

**Comment by scannillo** [comment]
Hello @drhaynes-goahead . Did you get a change to try this again using our updated [Carthage integration instructions here ](https://developers.braintreepayments.com/guides/3d-secure/migration/ios/v4#additional-setup-for-carthage) and using version 4.26.1 of the Braintree SDK?

**Comment by drhaynes** [comment]
Thanks for the updates. Am jumping back on this now, so will let you know how things go.

**Comment by drhaynes-goahead** [comment]
Can confirm, fixed with `braintree_ios` version 4.26.3 (`CardinalMobile` 2.1.4-2)
